src/Language/PureScript/Sugar/TypeClasses/Deriving.hs

```diff
@@ -600,6 +613,71 @@ deriveOrd1 ss =
     dataOrdCompare :: Expr
     dataOrdCompare = Var ss (Qualified (Just dataOrd) (Ident C.compare))
 
+deriveHashable :: forall m
+   . (MonadError MultipleErrors m, MonadSupply m)
+  => SourceSpan
+  -> ModuleName
+  -> SynonymMap
+  -> [Declaration]
+  -> ProperName 'TypeName
+  -> m [Declaration]
+deriveHashable ss mn syns ds tyConNm = do
+  tyCon <- findTypeDecl ss tyConNm ds
+  hashFun <- mkHashFunction tyCon
+  return [ ValueDecl (ss, []) (Ident C.hash) Public [] (unguarded hashFun) ]
+  where
+    mkHashFunction :: Declaration -> m Expr
+    mkHashFunction (DataDeclaration (ss', _) _ _ _ args) = do
+      x <- freshIdent "x"
+      lamCase ss' x <$> (concat <$> mapM mkCases (zip args [0..]))
+    mkHashFunction _ = internalError "mkHashFunction: expected DataDeclaration"
+
+    mkCases :: ((ProperName 'ConstructorName, [Type]), Int) -> m [CaseAlternative]
+    mkCases ((ctorName, tys), nth) = do
+      xs <- replicateM (length tys) (freshIdent "x")
+      let binder = ConstructorBinder ss (Qualified (Just mn) ctorName) (map (VarBinder ss) xs)
+      tys' <- mapM (replaceAllTypeSynonymsM syns) tys
+
+      -- collect all record and constructor fields
+      let es = collectFields (zip (map var xs) tys')
+      -- make recursive calls to `hash`
+      let hashEs = map (App (Var ss (Qualified (Just dataHashable) (Ident C.hash)))) es
+      hs <- replicateM (length es) (freshIdent "h")
+      -- unwrap all of the `Hash` newtypes
+      let theCase = Case hashEs
+                    [ CaseAlternative (map hashBinder hs)
+                      -- combine hashes, then wrap in newtype
+                      (unguarded (wrapHash (foldl' combiningFunction (initialHash nth) (map var hs)))) ]
+
+      return [CaseAlternative [binder] (unguarded theCase)]
+
+    collectFields [] = []
+    collectFields ((e,ty):rest)
+      | Just rec <- objectType ty
+      , Just fields <- decomposeRec rec
+          = collectFields (map (\((Label str), fieldTy) -> (Accessor str e, fieldTy)) fields) ++ collectFields rest
+      | otherwise = e:collectFields rest
+
+    var = Var ss . Qualified Nothing
+
+    dataHashableHashNewtypeConstructor = (Qualified (Just dataHashable) (ProperName "Hash"))
+
+    wrapHash = App (Constructor ss dataHashableHashNewtypeConstructor)
+
+    hashBinder :: Ident -> Binder
+    hashBinder hx = ConstructorBinder ss dataHashableHashNewtypeConstructor [ VarBinder ss hx ]
+
+    -- acc * 31 + h -- for want of a better combining function
```

Comment:
What is the tradeoff here?

Reply:
This is the combining function used by Java for lists etc. (see https://docs.oracle.com/javase/7/docs/api/java/util/List.html#hashCode()). It has issues when used with strings, where certain length prefixes lead to collisions. I'm not sure how much of a problem that is in practice with arbitrary datatypes.

The Haskell hashable library uses `combine h1 h2 = (h1 * 16777619) `xor` h2` (FNV-1, see https://en.wikipedia.org/wiki/Fowler%E2%80%93Noll%E2%80%93Vo_hash_function). I think the reason I didn't use this, is that I thought `xor` is not inlined like * and +, but it appears I'm wrong about that.

A better hash function would probably be something like murmur_32, which is used by Google's Guava library as the default for short hashes. The main problem is that it's quite a bit more code, which is a pain to generate in deriving. It's also slower to compute.

We could probably do even more clever things in the future. For small types we could find (minimal?) perfect hash functions, for example.

Related: I'm not at all certain that the seeds [0..] are the best we can do.